app/controllers/course/assessment/submission/answer/text_response/text_response_controller.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+class Course::Assessment::Submission::Answer::TextResponse::TextResponseController < \
+  Course::Assessment::Submission::Answer::TextResponse::Controller
+  load_resource :actable, class: Course::Assessment::Answer::TextResponse.name,
+                          singleton: true, through: :answer
+  before_action :set_text_response_answer
+
+  helper Course::Assessment::Submission::SubmissionsHelper.name.sub(/Helper$/, '')
+
+  def upload_files
+    success = @text_response_answer.class.transaction do
+      raise ActiveRecord::Rollback unless update_answer_files(upload_files_params)
+
+      true
+    end
+
+    if success
+      render @text_response_answer.answer
+    else
+      @text_response_answer.errors.messages.each do |attribute, message|
+        @text_response_answer.answer.errors.add(attribute, message)
+      end
+      render json: { errors: @text_response_answer.answer.errors.messages }, status: :bad_request
+    end
+  end
+
+  def delete_file
+    attachment_reference = AttachmentReference.find(delete_file_params[:attachment_id])
+    answer = @text_response_answer.acting_as
+
+    success = @text_response_answer.class.transaction do
+      answer.update(last_session_id: session.id, client_version: delete_file_params[:client_version])
```

Comment:
What happens when update fails? Will the transaction be triggered?

Reply:
I made a mistake here. Have just checked and the ActiveRecord::Rollback needs to be invoked manually also here. I have refactored it accordingly, together with the error message that will be shown in the frontend also has been fixed

Reply:
no wrong! you just use `update!` and there will be an error, then it will be rolled back automatically. Please read this again... https://medium.com/@kristenrogers.kr75/rails-transactions-the-complete-guide-7b5c00c604fc

Reply:
ah sorry my bad.. it's much simpler to use `update!`, done for this.